src/core/DatetimeClasses.ts

```diff
@@ -0,0 +1,475 @@
+     * @param offsetSeconds Must be between -64800-64800 (-+18:00)
+     */
+    static fromDate(date: Date, offsetSeconds: number): HzOffsetDateTime {
+        if (!(date instanceof Date) || isNaN(date.getTime())) {
+            throw new RangeError('Invalid date.');
+        }
+        if (!Number.isInteger(offsetSeconds) || !(offsetSeconds >= -64800 && offsetSeconds <= 64800)) {
+            throw new RangeError('Offset seconds can be between -64800(-18:00) and 64800(+18:00).');
+        }
+        return new HzOffsetDateTime(HzLocalDateTime.fromDate(date), offsetSeconds);
+    }
+
+    /**
+     * Constructs a new instance from ISO 8601 string.
+     * @param isoString ISO 8601 string with timezone. If timezone is omitted, UTC is assumed
+     */
+    static fromString(isoString: string): HzOffsetDateTime {
+        if (typeof isoString !== 'string') {
+            throw new RangeError('String expected');
+        }
+        const indexOfFirstMatch = isoString.search(HzOffsetDateTime.timezoneRegex);
+        const split = isoString.split(isoString[indexOfFirstMatch]);
+        let offsetSeconds;
+        if (split.length !== 2) {
+            throw new RangeError('Invalid format');
+        }
+        if (indexOfFirstMatch === -1) {
+            offsetSeconds = 0;
+        } else {
+            offsetSeconds = getOffsetSecondsFromTimezoneString(isoString[indexOfFirstMatch] + split[1]);
+        }
+        return new HzOffsetDateTime(HzLocalDateTime.fromString(split[0]), offsetSeconds);
+    }
+
+    /**
+     * Returns this offset datetime as Date. Note that the timezone information is not stored in Date objects and you
+     * effectively get a timestamp.(an instance in time without timezone)
+     */
+    asDate(): Date {
+        return new Date(this.hzLocalDateTime.asDate().getTime() - this.offsetSeconds * 1000);
+    }
+
+    /**
+     * Returns ISO 8601 string with timezone of this instance.
+     * @returns A string in the format yyyy-mm-ddTHH-mm-ss.SSS(Z | (+|-)HH:mm)
+     * Timezone is denoted either with `Z` or timezone string like +-HH:mm
+     */
+    toString(): string {
+        const timezoneOffsetString = getTimezoneOffsetFromSeconds(this.offsetSeconds);
+        return this.hzLocalDateTime.toString() + timezoneOffsetString;
+    }
+}
+
+
+interface HzLocalDateInterface {
+    (year: number, month: number, date: number): HzLocalDate;
+
+    new(year: number, month: number, date: number): HzLocalDate;
+}
+
+function _LocalDate(year: number, month: number, date: number): HzLocalDate {
+    return HzLocalDate._new(year, month, date);
+}
+
+/**
+ * Constructor function, can be invoked with or without new keyword.
+ *
+ * @param year Must be between -999999999-999999999
+ * @param month Must be between 1-12
+ * @param date Must be between 1-31 depending on year and month
+ * @throws RangeError if any of the arguments are invalid
+ */
+export const LocalDate: HzLocalDateInterface = <HzLocalDateInterface>_LocalDate;
```

Comment:
Why do we have those? How it is different than the constructor we have? If you want to provide a static factory method, then make it public on the class itself.

I don't think this abstraction is useful

Similar for others

Reply:
I explained above. The normal constructor cannot be used without `new`. I think static factory methods are a little bit harder to use given that some date/time classes accept each other, the code needed to build a HzOffsetDateTime will be too long.

Reply:
I am not sure about that. So, with the constructor we have
```js
new HzLocalDateTime(new HzLocalDate(2000, 1, 1), new HzLocalTime(12, 12, 12))
```
with factory methods we have
```js
HzLocalDateTime.from(HzLocalDate.from(200, 1, 1), HzLocalTime.from(12, 12, 12))
```
They seem the same to me. Note that, we can even make the constructor private and just provide the static methods

Reply:
In the current(constructor method export) version it is like this:

```js
OffsetDateTime(LocalDateTime(LocalDate(2000, 1, 1), LocalTime(12, 12, 12)), 1000)
```

I removed `hz` prefix from constructors. The ones with Hz are class names. API docs include both constructor and class documentation.

 Class constructor is private with the current approach

But I realized that maybe we can export a static method of the class, instead of defining a new function. If it works without new, it is ok.

Note that one of the most popular BigDecimal library big.js uses a similar constructor approach that can be used with or without the `new` keyword. Another benefit of exporting a named constant is that you can rename the constructor to a shorter variable while keeping the class name long to make it more descriptive.

Reply:
I still don't think the constructor without `new` is necessary. I mean, the argument that `big.js` does it like that is not enough IMHO. I could probably find dozens of popular libraries that do not provide such things as well.

BTW, I am OK with omitting Hz suffix from the classes and naming them to LocalTime, etc.

Reply:
I see no problem since my approach is wider, it can be also used with new. However, to be consistent with other classes we provide, I will remove the newless approach and user will use consturtors for construction. I will rename the classes to their names without HZ. I will add a from static method to localdatetime and offsetdatetime that takes all the input at once, instead of nested class arguments.

Reply:
I could be Ok with something like `BigDecimal()` and `new BigDecimal()`, but I believe the name should be the same as the class

Reply:
Maybe something like this, from TypeScript
```ts
interface Boolean {
    /** Returns the primitive value of the specified object. */
    valueOf(): boolean;
}

interface BooleanConstructor {
    new(value?: any): Boolean;
    <T>(value?: T): boolean;
    readonly prototype: Boolean;
}

declare var Boolean: BooleanConstructor;
```

Reply:
If you are ok with newless BigDecimal, then why not with datetime classes?

> I still don't think the constructor without new is necessary. I mean, the argument that big.js does it like that is not enough IMHO. I could probably find dozens of popular libraries that do not provide such things as well.

Also finding exported traditional classes in the popular libraries are very hard. I struggled to find one, I don't remember which one I found. Usually libraries(express, fastify, axios ) just export a function that is called and it is ready.

Reply:
I might be OK with them also. I just want to have just one name for them to avoid any confusion

Reply:
I feel like I can try to follow big.js's approach it exports one name that can be used both for type and constructor. In this case we will probably need to stop defining class and use classic function approach.

see https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/big.js/index.d.ts#L343

Reply:
Is such thing a pattern to avoid in TypeScript codebases? I don't know much about conventions in TypeScript but I assume mixing the `class` and `functions` in the same codebase does not look good. That's just my opinion though, if it is OK, feel free to do it

Reply:
How about the new approach? An interface can be exported with same name of another variable.

commits:
https://github.com/hazelcast/hazelcast-nodejs-client/pull/963/commits/242eaefd055eb9ba790f5fde55383bceea2e75fc
https://github.com/hazelcast/hazelcast-nodejs-client/pull/963/commits/38799e33f607866b463e9f9cfcb6bf5666533f34
https://github.com/hazelcast/hazelcast-nodejs-client/pull/963/commits/fc4370fde3cf95602006a0c0068935e25e655020

A single name is exported, can be used with or without new. The intellisense works perfectly now, in order it to happen, for intellij ide I needed to move constructor method above interface. I think it picks the first one it finds when there are more than one exports with same name.

Reply:
What do you think? Should we remove this newless approach altogether since it makes the code complex? If it will stay we may consider renaming BigDecimal to Big.

Reply:
Well, I am not sure about this approach because I don't know how that fits into the best practices of TypeScript. We now using the old way of doing things instead of the `class`es, right? Is this discouraged?

Reply:
I think not discouraged, but not very commonly used by libraries. Anyways, I will revert to the classic approach.

Reply:
See https://github.com/hazelcast/hazelcast-nodejs-client/commit/319f18ec1649fa95f3061de4cf8de10f780ee32b for reverted constructor change for BigDecimal and Date/Time

Added `static new()` to OffsetDatetime and LocalDateTime for easier construction. 

Also changed errors to TypeError as you suggested.